sdk/src/Core/Amazon.Runtime/Internal/Transform/StreamingUtf8JsonReader.cs

```diff
@@ -0,0 +1,134 @@
+        private void HandleUtf8Bom(ref byte[] buffer)
+        {
+            int utf8BomLength = JsonConstants.Utf8Bom.Length;
+            Debug.Assert(buffer.Length >= utf8BomLength);
+
+            int bytesRead = _stream.Read(buffer, 0, buffer.Length);
+            int start = 0;
+            if (buffer.AsSpan().StartsWith(JsonConstants.Utf8Bom))
+            {
+                // this additional copy only happens once if a UTF8 bom exists.
+                start += utf8BomLength;
+                bytesRead -= utf8BomLength;
+                Array.Copy(buffer, start, buffer, 0, bytesRead);
+            }
+            Array.Clear(buffer, bytesRead, buffer.Length - bytesRead);
+            _reader = new Utf8JsonReader(buffer, isFinalBlock: bytesRead == 0, default);
+        }
+
+        // Custom delegate to handle ref parameters
+        public delegate void RefAction(ref Utf8JsonReader reader);
+
+        /// <summary>
+        /// Method to allow passing the private _reader as ref. This is a hacky way to get around the error
+        /// "a property or indexer may not be passed as an out or ref parameter". Use this when you need to pass
+        /// the Utf8JsonReader to a method that requires a ref parameter. <see cref="JsonUnmarshallerContext.ToJsonDocument(ref StreamingUtf8JsonReader)"/>
+        /// for usage example.
+        /// </summary>
+        /// <param name="action">The custom delegate</param>
+
+        public void PassReaderByRef(RefAction action)
+        {
+            action(ref _reader);
+        }
+
+        /// <summary>
+        /// The Reads data from the buffer and fetches more data from the stream to fill the buffer it more data exists.
+        /// </summary>
+        /// <returns>true if there is more data, false otherwise.</returns>
+        public bool Read()
+        {
+            bool hasMoreData = _reader.Read();
+            if (!hasMoreData)
+            {
+                GetMoreBytesFromStream(_stream, ref _buffer, ref _reader);
+                hasMoreData = _reader.Read();
+            }
+
+            return hasMoreData;
+
+        }
+
+        private static void GetMoreBytesFromStream(Stream stream, ref byte[] buffer, ref Utf8JsonReader reader)
+        {
+            int bytesRead;
+            if (reader.BytesConsumed < buffer.Length)
+            {
+                ReadOnlySpan<byte> leftover = buffer.AsSpan((int)reader.BytesConsumed);
+
+                if (leftover.Length == buffer.Length)
+                {
+                    Array.Resize(ref buffer, buffer.Length * 2);
+                }
+
+                leftover.CopyTo(buffer);
+                bytesRead = stream.Read(buffer, leftover.Length, buffer.Length - leftover.Length);
+            }
+            else
+            {
+                bytesRead = stream.Read(buffer, 0, buffer.Length);
+            }
+            if (bytesRead == 0)
```

Comment:
should the method return here, I worry that the follow line sets the reader using the `buffer` that is being returned to the pool

Reply:
ah, yeah good suggestion the reader would be using a buffer returned to the pool in that scenario. If I return there though, it will return to the `Read()` method and the `_reader` won't have a buffer to work with, and I don't want to set buffer to null. I think instead I'll just refactor a bit and add something like the following in the `Read()` method
```
            if (!hasMoreData)
            {
                GetMoreBytesFromStream(_stream, ref _buffer, ref _reader);
                hasMoreData = _reader.Read();
                if (_reader.IsFinalBlock)
                   ArrayPool<byte>.Shared.Return(_buffer);
            }
```

Reply:
while fixing a bug in the streamingutf8 reader I decided to remove the pooling due to multiple resizings of the array. I put a justification in a comment, but basically the byte array returned by `Array.Resize` isn't owned by the pool so keeping track of that didn't seem worth the Renting + copying I would have to do in order to do that

Reply:
flipped back to pooling